README.md
# 🚀 Iron Hero Web Edition 0.1-Beta 🚀

![Screenshot](screenshots/screenshot.png)

Iron Hero is an exciting game where you can take control of a spaceship, battle alien foes, and conquer the cosmos. Please note that this game is still in development.

## How to Run the Game

To play the Iron Hero, follow these steps:

1. Clone the repository to your local machine:

   ```sh
   git clone https://github.com/jeremystevens/iron-hero.git

2. Navigate to the project directory:
   ```sh
   cd iron-hero

3. Run the game by Opening the following file in your web browser:
    ```sh
    open index.html
  

## Gameplay

- Use the **keyboard arrows** to move your spaceship:
  - **Up arrow**: Move up
  - **Down arrow**: Move down
  - **Left arrow**: Move left
  - **Right arrow**: Move right

- Press the **Spacebar** to shoot at your enemies and defend your spaceship.

## Contributing

If you'd like to contribute to this project, feel free to open an issue or submit a pull request. We welcome your contributions!

## License

This project is licensed under the MIT License. See the [LICENSE.md](LICENSE.md) file for details.


CHANGELOG.MD
# DEV CHANGELOG


## Dev Notes: Rev 0.2.3

- Added Player Class Moved all player attributes to Player Class. 
- Organized code to be more readable

## Dev Notes: Rev 0.2.2

- Added border around the game canvas to indicate the boundaries of the game window.
- Added version text to the HTML page, displaying it in the bottom right corner of the canvas.

## Dev Notes: Rev 0.2.1

- Modified the layout to display the Level first and the Score to the right of it.


## Dev Notes: Rev 0.2

- Modified the player lives indicator to display on the same line.
- Adjusted the size of the player sprite lives to make them smaller.
- Positioned the player sprite lives to the right of the "Lives:" text without overlapping by using a fixed xOffset.
- Added score and level display to the game interface.


## Dev Notes: Rev 0.1

- Initial Game setup (folder structure, assets, etc.).
- Added Title Screen.
- Added Title Music & Copyright info at the bottom of the screen.
- Added Player & Bullet Mechanics 
. 


mobile/nipple.js
var s = function(sel) {
  return document.querySelector(sel);
};
var sId = function(sel) {
  return document.getElementById(sel);
};
var removeClass = function(el, clss) {
  el.className = el.className.replace(new RegExp('\\b' + clss + ' ?\\b', 'g'), '');
}
var joysticks = {
  dynamic: {
    zone: s('.zone.dynamic'),
    color: 'blue',
    multitouch: true
  },
  semi: {
    zone: s('.zone.semi'),
    mode: 'semi',
    catchDistance: 150,
    color: 'white'
  },
  static: {
    zone: s('.zone.static'),
    mode: 'static',
    position: {
      left: '50%',
      top: '50%'
    },
    color: 'red'
  }
};
var joystick;

// Get debug elements and map them
var elDebug = sId('debug');
var elDump = elDebug.querySelector('.dump');
var els = {
  position: {
    x: elDebug.querySelector('.position .x .data'),
    y: elDebug.querySelector('.position .y .data')
  },
  force: elDebug.querySelector('.force .data'),
  pressure: elDebug.querySelector('.pressure .data'),
  distance: elDebug.querySelector('.distance .data'),
  angle: {
    radian: elDebug.querySelector('.angle .radian .data'),
    degree: elDebug.querySelector('.angle .degree .data')
  },
  direction: {
    x: elDebug.querySelector('.direction .x .data'),
    y: elDebug.querySelector('.direction .y .data'),
    angle: elDebug.querySelector('.direction .angle .data')
  }
};


var timeoutCreate;
function createThrottle (evt) {
  clearTimeout(timeoutCreate);
  timeoutCreate = setTimeout(() => {
    createNipple(evt);
  }, 100);
}

sId('buttons').onclick = sId('buttons').ontouchend = createThrottle;

createNipple('dynamic');

function bindNipple() {
  joystick.on('start end', function(evt, data) {
    dump(evt.type);
    debug(data);
  }).on('move', function(evt, data) {
    debug(data);
  }).on('dir:up plain:up dir:left plain:left dir:down ' +
        'plain:down dir:right plain:right',
        function(evt, data) {
    dump(evt.type);
  }
       ).on('pressure', function(evt, data) {
    debug({
      pressure: data
    });
  });
}

function createNipple(evt) {
  var type = typeof evt === 'string' ?
      evt : evt.target.getAttribute('data-type');
  if (joystick) {
    joystick.destroy();
  }
  removeClass(s('.zone.active'), 'active');
  removeClass(s('.button.active'), 'active');
  removeClass(s('.highlight.active'), 'active');
  s('.highlight.' + type).className += ' active';
  s('.button.' + type).className += ' active';
  s('.zone.' + type).className += ' active';
  joystick = nipplejs.create(joysticks[type]);
  console.log(joystick, nipplejs);
  bindNipple();
}

// Print data into elements
function debug(obj) {
  function parseObj(sub, el) {
    for (var i in sub) {
      if (typeof sub[i] === 'object' && el) {
        parseObj(sub[i], el[i]);
      } else if (el && el[i]) {
        el[i].innerHTML = sub[i];
      }
    } 


game.js
window.onload = function() {
    const canvas = document.getElementById('gameCanvas');
    const ctx = canvas.getContext('2d');

    // Display the version number on the title screen
    const versionNumber = "0.2.3";
    // Update the version text
    document.getElementById('version').textContent = versionNumber;

    // Title music
    let titleMusic = new Audio('audio/menu.wav');
    titleMusic.loop = true;

    // Game music
    let gameMusic = new Audio('audio/game.mp3');
    gameMusic.loop = true;

    // Variables to track score and level
    let score = 0;
    let level = 1;

    // Variable to track player lives
    let lives = 3; // Start with 3 lives

    // Set the size for the player lives indicator
    const playerLivesIndicatorSize = 30;
    const playerLivesSpriteSize = 20;

    // Function to draw player lives using the player sprite
    function drawPlayerLives() {
        ctx.fillStyle = 'white';
        ctx.font = '27px Arial';

        // Display "Lives:" text
        ctx.fillText('Lives: ', 10, 30);

        // Fixed xOffset for each player sprite life
        const xOffset = 90;
        const yoffset = 10;
        // Draw scaled-down player sprites for each remaining life
        for (let i = 0; i < lives; i++) {
            // Calculate the position for each player sprite life
            const xPos = xOffset + i * (playerLivesSpriteSize + 5);
            // Draw the player sprite for lives with the adjusted size
            ctx.drawImage(playerImage, xPos, 10, playerLivesSpriteSize, playerLivesSpriteSize);
        }
    }

    // Player variables
    let playerImage = new Image();
    playerImage.src = 'images/player_sprite.png';

    let bulletImage = new Image();
    bulletImage.src = 'images/bullet_sprite.png';

    // define bullet sound and position
    let bulletSound = new Audio('audio/blaster.mp3');
    let bulletAdjustX = 10;
    let bulletAdjustY = 19;

    // Set up player object
    class Player {
        constructor(x, y, width, height) {
            this.x = x;
            this.y = y;
            this.width = width;
            this.height = height;
            this.vx = 0;
            this.vy = 0;
            this.bullets = [];
            this.image = playerImage;
        }

        moveLeft() {
            this.vx = -5;
        }

        moveUp() {
            this.vy = -5;
        }

        moveRight() {
            this.vx = 5;
        }

        moveDown() {
            this.vy = 5;
        }

        stopMoving() {
            this.vx = 0;
            this.vy = 0;
        }

        shootBullet() {
            this.bullets.push({
                x: this.x + this.width / 2 - 5 + bulletAdjustX,
                y: this.y + bulletAdjustY,
                width: 10,
                height: 10,
            });
            bulletSound.play();
        }

        updateBullets() {
            for (let i = 0; i < this.bullets.length; i++) {
                let bullet = this.bullets[i];
                bullet.x += 10; // Replace with bullet's speed
            }
        }
    }

    let player = new Player(canvas.width / 2, canvas.height - 100, 50, 50);

    // Title screen image
    let titleScreen = new Image();
    titleScreen.onload = function() {
        ctx.drawImage(titleScreen, 0, 0, canvas.width, canvas.height);
        drawButtons();
    };
    titleScreen.src = 'images/title_screen_image.png';

    // Button dimensions and positions
    const buttonWidth = 200;
    const buttonHeight = 50;
    const buttonColor = '#004A7F';
    const buttonTextColor = '#FFFFFF';
    const font = '36px Arial';

    // Function to draw score and level
    function drawGameInfo() {
        ctx.fillStyle = 'white';
        ctx.font = '24px Arial';

        // Display "Level:" text
        ctx.fillText('Level: ' + level, canvas.width - 250, 30);

        // Display "Score:" text
        ctx.fillText('Score: ' + score, canvas.width - 150, 30);

        drawPlayerLives();
    }

    // Draw buttons and text
    function drawButtons() {
        ctx.fillStyle = buttonColor;

        // Play button
        const playButtonRect = {
            x: canvas.width / 2 - buttonWidth / 2,
            y: 250,
            width: buttonWidth,
            height: buttonHeight
        };
        ctx.fillRect(playButtonRect.x, playButtonRect.y, playButtonRect.width, playButtonRect.height);

        // Settings button
        const settingsButtonRect = {
            x: canvas.width / 2 - buttonWidth / 2,
            y: 320,
            width: buttonWidth,
            height: buttonHeight
        };
        ctx.fillRect(settingsButtonRect.x, settingsButtonRect.y, settingsButtonRect.width, settingsButtonRect.height);

        // Draw button text
        ctx.fillStyle = buttonTextColor;
        ctx.font = font;
        drawCenteredText("Play", playButtonRect);
        drawCenteredText("Settings", settingsButtonRect);

        // Draw copyright text
        ctx.font = "16px Arial";
        ctx.fillStyle = 'white';
        const copyrightText = "© 2023 Jeremy Stevens";
        const textWidth = ctx.measureText(copyrightText).width;
        const textX = canvas.width - textWidth - 10; // 10 pixels from the right edge
        const textY = canvas.height - 10; // 10 pixels from the bottom edge
        ctx.fillText(copyrightText, textX, textY);
    }

    function drawCenteredText(text, rect) {
        const textWidth = ctx.measureText(text).width;
        const textX = rect.x + (rect.width - textWidth) / 2;
        const textY = rect.y + (rect.height + 30) / 2;
        ctx.fillText(text, textX, textY);
    }

    // Add event listener for canvas click
    canvas.addEventListener('click', function(event) {
        const rect = canvas.getBoundingClientRect();
        const x = event.clientX - rect.left;
        const y = event.clientY - rect.top;

        // Play button
        const playButtonRect = {
            x: canvas.width / 2 - buttonWidth / 2,
            y: 250,
            width: buttonWidth,
            height: buttonHeight
        };
        if (
            x > playButtonRect.x &&
            x < playButtonRect.x + playButtonRect.width &&
            y > playButtonRect.y &&
            y < playButtonRect.y + playButtonRect.height
        ) {
            titleMusic.pause(); // Pause title music
            startCountdown();
        }

        // Settings button
        const settingsButtonRect = {
            x: canvas.width / 2 - buttonWidth / 2,
            y: 320,
            width: buttonWidth,
            height: buttonHeight
        };
        if (
            x > settingsButtonRect.x &&
            x < settingsButtonRect.x + settingsButtonRect.width &&
            y > settingsButtonRect.y &&
            y < settingsButtonRect.y + settingsButtonRect.height
        ) {
            // handle settings button click TODO - add settings screen
            console.log("Settings button clicked");
        }
    });

    // Countdown logic
    let countdownValue = 5; // Starting value of the countdown

    function startCountdown() {
        if (countdownValue > 0) {
            playCountdownSound(countdownValue);
            drawCountdownNumber(countdownValue);
            countdownValue--;

            setTimeout(startCountdown, 1000); // Wait 1 second before the next number
        } else {
            // Countdown finished, proceed with starting the game
            startGame();
        }
    }

    function playCountdownSound(number) {
        let sound = new Audio('audio/' + number + ".mp3");
        sound.play();
    }

    function drawCountdownNumber(number) {
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        const bigFont = "120px Arial";
        ctx.font = bigFont;
        ctx.fillStyle = 'white';
        const text = number.toString();
        const textWidth = ctx.measureText(text).width;
        ctx.fillText(text, (canvas.width - textWidth) / 2, canvas.height / 2);
    }

    // Scrolling background
    let bgImage = new Image();
    bgImage.src = "images/background.png";
    let x1 = 0;
    let x2 = canvas.width;
    let speed = 17;

    // Game loop
    function gameLoop() {
        // Start playing the game music game.wav
        gameMusic.play();

        // Scroll the background
        x1 -= speed;
        x2 -= speed;

        if (x1 < -canvas.width) {
            x1 = canvas.width;
        }
        if (x2 < -canvas.width) {
            x2 = canvas.width;
        }

        // Draw the images
        ctx.drawImage(bgImage, x1, 0, canvas.width, canvas.height);
        ctx.drawImage(bgImage, x2, 0, canvas.width, canvas.height);

        // Update player position
        player.x += player.vx;
        player.y += player.vy;

        // Check bounds
        if (player.x < 0) player.x = 0;
        if (player.y < 0) player.y = 0;
        if (player.x + player.width > canvas.width) player.x = canvas.width - player.width;
        if (player.y + player.height > canvas.height) player.y = canvas.height - player.height;

        // Draw the player
        ctx.drawImage(player.image, player.x, player.y, player.width, player.height);

        // Draw the bullets
        for (let i = 0; i < player.bullets.length; i++) {
            let bullet = player.bullets[i];
            ctx.drawImage(bulletImage, bullet.x, bullet.y, bullet.width, bullet.height);
        }

        // Update bullet positions
        player.updateBullets();

        // Remove bullets that are off the screen
        player.bullets = player.bullets.filter(function(bullet) {
            return bullet.x < canvas.width;
        });

        // Draw game info (score and level)
        drawGameInfo();

        // Request next frame
        requestAnimationFrame(gameLoop);
    }

    // Start the game loop here
    function startGame() {
        // Start game loop
        requestAnimationFrame(gameLoop);
    }

    // Update player movement
    function updatePlayerMovement() {
        // Initialize velocity
        player.vx = 0;
        player.vy = 0;
    
        if (keys[37]) { // Left arrow
            player.vx = -5;
        }
        if (keys[38]) { // Up arrow
            player.vy = -5;
        }
        if (keys[39]) { // Right arrow
            player.vx = 5;
        }
        if (keys[40]) { // Down arrow
            player.vy = 5;
        }
    }

    // Handle keyboard input
    let keys = {};

window.addEventListener('keydown', function(event) {
    keys[event.keyCode] = true;
    if (event.keyCode === 32) { // Spacebar
        player.shootBullet();
    }
    updatePlayerMovement();
});

window.addEventListener('keyup', function(event) {
    keys[event.keyCode] = false;
    updatePlayerMovement();
});

    // Create an Audio object
    let menuMusic = new Audio('audio/menu.wav');

    // Play the music when the page is loaded
    window.addEventListener('load', function() {
        menuMusic.play();
    });
}; // end of file
